Mathlib/Data/Fin/Basic.lean

```diff
@@ -597,147 +615,264 @@ end Order
   rfl
 #align fin.mk_one Fin.mk_one
 
-instance {n : ℕ} : Nontrivial (Fin (n + 2)) :=
-  ⟨⟨0, 1, by decide⟩⟩
+instance nontrivial {n : ℕ} : Nontrivial (Fin (n + 2)) where
+  exists_pair_ne := ⟨0, 1, (ne_iff_vne 0 1).mpr (by simp only [val_one, val_zero])⟩
 
 theorem nontrivial_iff_two_le : Nontrivial (Fin n) ↔ 2 ≤ n := by
-  rcases n with (_ | _ | n) <;> simp [Fin.nontrivial, not_nontrivial, Nat.succ_le_iff]
+  rcases n with (_ | _ | n) <;>
+  simp [←Nat.one_eq_succ_zero, Fin.nontrivial, not_nontrivial, Nat.succ_le_iff]
+-- porting note: here and in the next lemma, had to use `←Nat.one_eq_succ_zero`.
 #align fin.nontrivial_iff_two_le Fin.nontrivial_iff_two_le
 
 theorem subsingleton_iff_le_one : Subsingleton (Fin n) ↔ n ≤ 1 := by
-  rcases n with (_ | _ | n) <;> simp [IsEmpty.subsingleton, Unique.subsingleton, not_subsingleton]
+  rcases n with (_ | _ | n) <;>
+  simp [IsEmpty.instSubsingleton, Unique.instSubsingleton, ←Nat.one_eq_succ_zero, not_subsingleton]
 #align fin.subsingleton_iff_le_one Fin.subsingleton_iff_le_one
 
 section Monoid
 
-@[simp]
-protected theorem add_zero [NeZero n] (k : Fin n) : k + 0 = k := by
+--Porting note: removing `simp`, `simp` can prove it with AddCommMonoid instance
+protected theorem add_zero (k : Fin (n + 1)) : k + 0 = k := by
   simp [eq_iff_veq, add_def, mod_eq_of_lt (is_lt k)]
 #align fin.add_zero Fin.add_zero
 
-@[simp]
-protected theorem zero_add [NeZero n] (k : Fin n) : (0 : Fin n) + k = k := by
+--Porting note: removing `simp`, `simp` can prove it with AddCommMonoid instance
+protected theorem zero_add (k : Fin (n + 1)) : (0 : Fin (n + 1)) + k = k := by
   simp [eq_iff_veq, add_def, mod_eq_of_lt (is_lt k)]
 #align fin.zero_add Fin.zero_add
 
-instance addCommMonoid (n : ℕ) [NeZero n] : AddCommMonoid (Fin n)
-    where
-  add := (· + ·)
-  add_assoc := by simp [eq_iff_veq, add_def, add_assoc]
-  zero := 0
-  zero_add := Fin.zero_add
-  add_zero := Fin.add_zero
-  add_comm := by simp [eq_iff_veq, add_def, add_comm]
+@[to_additive_fixed_numeral]
+instance [NeZero n] : OfNat (Fin n) a where
+  ofNat := Fin.ofNat' a Fin.size_positive'
+
+instance (n) : AddCommSemigroup (Fin n) where
+  add_assoc _ _ _ := by
+    apply Fin.eq_of_val_eq
+    simp only [Fin.add_def, Nat.mod_add_mod, Nat.add_mod_mod, Nat.add_assoc]
+  add_comm _ _ := by
+    apply Fin.eq_of_val_eq
+    simp only [Fin.add_def, Nat.add_comm]
+
+instance (n) : CommSemigroup (Fin n) where
+  mul_assoc a b c := by
+    apply Fin.eq_of_val_eq
+    simp only [Fin.mul_def]
+    generalize lhs : ((a.val * b.val) % n * c.val) % n = l
+    generalize rhs : a.val * (b.val * c.val % n) % n = r
+    rw [← Nat.mod_eq_of_lt c.isLt, (Nat.mul_mod (a.val * b.val) c.val n).symm] at lhs
+    rw [← Nat.mod_eq_of_lt a.isLt, (Nat.mul_mod a.val (b.val * c.val) n).symm,
+        ← Nat.mul_assoc] at rhs
+    rw [← lhs, ← rhs]
+  mul_comm (a b : Fin n) : a * b = b * a := by
+    apply Fin.eq_of_val_eq; simp only [Fin.mul_def, Nat.mul_comm]
```

Comment:
Where did this come from? In lean3 this instance is in data/zmod/basic, so I don't think we should have it in this file.